templates/al2023/runtime/rootfs/usr/bin/configure-multicard-interfaces.sh

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env bash
+
+# EKS already forces amazon-ec2-net-utils to manage the primary ENI only (0/0):
+# https://github.com/awslabs/amazon-eks-ami/pull/1738
+#
+# When VPC CNI is installed it creates another ENI from network card with index 0 (0/1).
+# VPC CNI will continue adding additional ENIs until max number of interfaces for the intance
+# is reached (0/2, 0/3, etc).
+#
+# This script configures IP and routing for Elastic Network Interfaces that are part of
+# non-zero indexed EC2 network cards (1/0, 1/1, 2/1, 3/1, etc). The way we find out whether
+# we need to configure the interface is by checking IMDS:
+# /latest/meta-data/network/interfaces/macs/${mac_address}/network-card/
+# This script will skip any interfaces that are part of the 0 indexed card.
```

Comment:
A+++++ comment, thx.

I'm wondering if we can leverage systemd-networkd here instead of setting all the routes up ourselves. We modify the default networkd config so that it only targets the primary ENI (by using `PermanentMACAddress`). Can we add more MAC's there for the other interfaces on other cards? Or create additional networkd configs for those cards?

Reply:
If we add the mac addresses to the current drop-in file, the default routes for them will end up in the main route table, which I think it's not a good idea.

If we want to leverage `systemd-networkd` we'll have to create separate `.network` file for the non zero cards populate the mac addresses and configure the policy routing. I was thinking about that too, but figured it is probably easier/clear to setup everything with `iproute` package commands.  I'm fine with changing direction, also asked somebody more familiar with networking to take a look, let's see what their take will be too :)

Reply:
> it is probably easier/clear to setup everything with iproute package commands

I would much rather defer to networkd 😂  we have to phone a friend every time we touch this script because it's pretty esoteric stuff

Reply:
+1 to seeing what we can do to avoid adding this into our AMIs for AL2023. I think we need to figure out how amazon-ec2-net-utils can support management of just the primary ENI and let others (like VPC CNI) handle all secondary ENI management.

We're also going through this much effort to bring up network interfaces that aren't on network card 0, but those interfaces aren't actually used by K8s pods in any case since we only assign IPs from network card 0 right? This is something we've got to chase with the VPC CNI to make it network card aware, but until then I wonder how much value this PR brings us other than consistency with our AL2 AMI.

Reply:
>+1 to seeing what we can do to avoid adding this into our AMIs for AL2023. I think we need to figure out how amazon-ec2-net-utils can support management of just the primary ENI and let others (like VPC CNI) handle all secondary ENI manageme

We already "tell" ec2-net-utils to manage the primary ENI, by using [drop-in](https://github.com/awslabs/amazon-eks-ami/pull/1738). When checking with VPC CNI people I was told that there is backlog item about handling non primary interfaces, but I didn't get anything more than that. I can ask further. 

>We're also going through this much effort to bring up network interfaces that aren't on network card 0, but those interfaces aren't actually used by K8s pods in any case since we only assign IPs from network card 0 right?

Potential use case is if someone runs the pod in the host network namespace and specifically binds their app to one of non zero card interfaces.

>I wonder how much value this PR brings us other than consistency with our AL2 AMI.

My main goal was parity with AL2 AMI.

Reply:
> +1 to seeing what we can do to avoid adding this into our AMIs for AL2023. I think we need to figure out how amazon-ec2-net-utils can support management of just the primary ENI and let others (like VPC CNI) handle all secondary ENI management

Making VPC CNI aware of network cards other than card 0 is going to be a much bigger change and currently not prioritized by the networking team. We went ahead with same approach as AL2 AMI to not block the AL2023 AMI on this.
cc: @jayanthvn on setting up routes via iproute vs using networkd by creating separate .network files

Reply:
The CNI plugin will not manage the ENI setup and routes for the primary ENIs on non-zero network cards, even if the CNI plugin manages the secondary ENIs on the non-zero network cards.